### DISCUSSION TITLE
quarto publish problem: can not start python3 kernel

### DISCUSSION THREAD

**Discussion by xinhuaChenW** [quarto-dev/quarto-cli#2597] [Q&A]
Could you please help?

When Quarto was publishing new '.qmd' containing python codes, it stucked at 'Starting python3 kernel'. It appears that Quarto can't start python kernel. 
figure 1. 
<img width="382" alt="image" src="https://user-images.githubusercontent.com/63830206/192121706-61c2ac84-c1e4-4811-8e2a-f37ebd35df8e.png">

However, when the python code cell is removed, the .qmd file can be published successfully. So, the problem should be correlated to the python kernel.

The path to the python kernel is right, however, Quarto just can't start the kernel. I have no idea what is going on. 

figure 2. the python interpreter pathway in setting.json file. 
<img width="477" alt="image" src="https://user-images.githubusercontent.com/63830206/192121030-376afdeb-2b7c-4ba8-9b58-98f57e91c10d.png">

figure 3. the pathway of the current running python interpreter in VSCode is exactly the same with the pathway in setting.json as in figure 1.
<img width="133" alt="image" src="https://user-images.githubusercontent.com/63830206/192121069-7804e98a-8072-4f44-b5d8-8e791a67c0c8.png">

And it makes me more confused that the python code in this '.qmd' file can be run if I click the 'Run Cell ' button and it gives correct outputs.  

figure 4 and 5. python code can give output correctly if the code cell is executed alone.
<img width="330" alt="image" src="https://user-images.githubusercontent.com/63830206/192121843-36d424f7-8473-4aa2-9314-994297b89d4f.png">

<img width="403" alt="image" src="https://user-images.githubusercontent.com/63830206/192121186-ae775323-be32-4111-8bd6-e6bcfd4cd8a4.png">

setting.json about python interpreter path is as following.
``` 
    "python.defaultInterpreterPath": "D:\\Anaconda\\python.exe",
    "editor.inlineSuggest.enabled": true,
    "python.condaPath": "D:\\Anaconda\\_conda.exe",
```

**Comment by eitsupi** [comment]
I don't think the quarto executed on the command line and vscode settings are relevant.

**Reply by xinhuaChenW** [reply]
Thank for your reply! 
I tried something esle according to a similar question: https://github.com/quarto-dev/quarto-cli/discussions/1372
I tried `quarto check`, the check was stucked at `Checking Jupyter engine render....`
<img width="263" alt="image" src="https://user-images.githubusercontent.com/63830206/192184887-b3fac177-f926-4b80-b6f3-0868f44cdc45.png">
Does that mean it has something to do with the jupyter engine?

**Reply by bvancil** [reply]
Quarto also hangs from the command line for me, even when trying to execute the relatively simple example at <https://quarto.org/docs/computations/python.html>. Quarto hangs at `Starting python3 kernel...`.

My `quarto check` hangs at the same place:

```
[>] Checking Quarto installation......OK
      Version: 1.1.251
      Path: C:\Users\bvancil\scoop\apps\quarto\current\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.10.5
      Path: C:/Users/bvancil/.virtualenvs/quarto-sas-example-Ugam9axL-python/Scripts/python.exe
      Jupyter: 4.11.1
      Kernels: julia-1.7, julia-1.8.2-1.8, sas, python3

(-) Checking Jupyter engine render....
```

**Reply by xinhuaChenW** [reply]
Thank you for your reply. Similar problem in Ubuntu 22.0 LTS environment. 😁

**Reply by bvancil** [reply]
I haven't tried this recently? @cxh76153264 , have you? Maybe we should submit this as an issue rather than a discussion.

**Reply by bvancil** [reply]
I figured out the problem, for me at least. I was inadvertently allowing prerelease versions of `jupyter-client` and `notebook` python packages. Once I pinned those to stable release versions, `quarto check jupyter` worked fine. This might be a heads-up to the Quarto devs that jupyter-client 8.0.0a breaks things.

**Reply by lindsay-fredrick** [reply]
For what it's worth, the current version still breaks things - I've pinned jupyter_client to 7.3.5 for now. jupyter-client 8.0.1 is no longer pre-release so I just found out today that it isn't playing nicely when I tried to render my github-page without a pinned version.

**Reply by cscheid** [reply]
@lindsay-forestell If you can produce a minimal case where this happens for you consistently, we'd appreciate a bug report, thank you.

**Reply by xinhuaChenW** [reply]
> I haven't tried this recently? @cxh76153264 , have you? Maybe we should submit this as an issue rather than a discussion.

Sorry to reply so late. bvancil. Glad to recieve a good news from you. Since I encountered that problem last time, I haven't used quoto for a long time. I've been busy with other more urgent things these days and haven't finished yet. I admire you for continuously trying to research this issue and finding a solution. I'm also interested because maybe I can use quarto in VScode happily again (cause no need to swicth to another IDE). When I have some free time, I'll try your method. Thank you for your reply, I really appreciate it!!

**Reply by xinhuaChenW** [reply]
@bvancil It is working well with jupyter-client 7.4.8, facinate!! Thank you !!

**Comment by benabel** [comment]
I also add to pin `jupyter_client==7.4.7` to build with quarto on a docker-image using ubuntu:rolling. If not quarto hanged on `...Starting jupyter kernel`

**Reply by cscheid** [reply]
If you use one of our latest prereleases (as of yesterday) then you should be able to make things work with jupyter_client 8+. Our very latest prerelease (being built just now) should work with _both_ versions.

**Reply by xinhuaChenW** [reply]
Thank you for your reply! Will give it a try later. Happy about using quarto again.

**Comment by Mouli04** [comment]
I too facing the same error. Any fix...?

**Reply by dragonstyle** [reply]
We just published Quarto 1.2.335 which should fix this issue.

https://quarto.org/docs/get-started/

**Comment by benabel** [comment]
I confirm this is fixed on my system with v1.3.159. Thank you.
